### PR TITLE
WIP: add crude config file support

### PR DESCRIPTION
**Pull request by granquet**
Signed-off-by: Guillaume Ranquet <ranquet.guillaume@gmail.com>

This is a very crude implementation of a config file support...
Pull request mainly for discussion.
if upstream is interested in the feature, I'll rework the patch to have something a bit more robust.

### CODE REVIEW

**Comment by Airblader** [comment]
> Pull request mainly for discussion.

Fair enough, but before we can start that I would ask you to explain a bit as to why you would like to see config files for i3lock as that's the main question to answer, right? :-)

One reason I personally could see is that it would allow shipping things like systemd unit files while leaving configuration to the user, although in this case we'd probably want to have behavior such as command line arguments always overriding the config file.

**Comment by granquet** [comment]
Oh... mainly because I'm lazy :)

I'm running dmenu and call i3lock from dmenu.
and as you probably know, dmenu doesn't handle aliases... which means I'm always locking my screen with that white default screen...

I could run i3lock from a shell prompt, but most of the times I'm ssh'ing somewhere or running inside a tmux session with DISPLAY unset (because I started the session from another machine using ssh usually).

I could also create a script and move it in my $PATH...

I like the config file better, it's easier to bring with me when I'm getting a new account on another machine.

this is already coded so that the config file is always overriden by cmdline arguments as the config_file is parsed before the call to getopt_long().

**Comment by ivanhercaz** [comment]
A configuration file to save time configuring another computer or make easier to have a global/general configuration would be very useful. @Airblader, would it be possible to apply this to i3lock?

I don't know very much about C, but if I am understanding fine the check done by Travis, the problem is in the format of the new code, specifically with the indentation. Am I right? I am going to open a review to point out these mistakes.

File: i3lock.c [hunk 1]
```diff
@@ -1004,105 +1004,148 @@ static void raise_loop(xcb_window_t window) {
...
+            break;
+        case 'b':
+            beep = true;
+            break;
+        case 'd':
+            fprintf(stderr, "DPMS support has been removed from i3lock. Please see the manpage i3lock(1).\n");
+            break;
+        case 'I': {
+                      fprintf(stderr, "Inactivity timeout only makes sense with DPMS, which was removed. Please see the manpage i3lock(1).\n");
+                      break;
```
Comment (ivanhercaz) [inline i3lock.c]: Indentation error, should be like in `case 'd':`.

File: i3lock.c [hunk 2]
```diff
@@ -1004,105 +1004,148 @@ static void raise_loop(xcb_window_t window) {
...
+                  }
+        case 'c': {
+                      char *arg = optarg;
```
Comment (ivanhercaz) [inline i3lock.c]: Indentation error, should be like in `case 'd':`.

File: i3lock.c [hunk 3]
```diff
@@ -1004,105 +1004,148 @@ static void raise_loop(xcb_window_t window) {
...
+                      if (arg[0] == '#')
+                          arg++;
+
+                      if (strlen(arg) != 6 || sscanf(arg, "%06[0-9a-fA-F]", color) != 1)
+                          errx(EXIT_FAILURE, "color is invalid, it must be given in 3-byte hexadecimal format: rrggbb");
+
+                      break;
+                  }
+        case 'u':
+                  unlock_indicator = false;
```
Comment (ivanhercaz) [inline i3lock.c]: Indentation error, should be like in `case 'd':`.

File: i3lock.c [hunk 4]
```diff
@@ -1004,105 +1004,148 @@ static void raise_loop(xcb_window_t window) {
...
+                  break;
+        case 'i':
+                  image_path = strdup(optarg);
```
Comment (ivanhercaz) [inline i3lock.c]: Indentation error, should be like in `case 'd':`.

File: i3lock.c [hunk 5]
```diff
@@ -1004,105 +1004,148 @@ static void raise_loop(xcb_window_t window) {
...
+                  break;
+        case 't':
+                  tile = true;
```
Comment (ivanhercaz) [inline i3lock.c]: Indentation error, should be like in `case 'd':`.

File: i3lock.c [hunk 6]
```diff
@@ -1004,105 +1004,148 @@ static void raise_loop(xcb_window_t window) {
...
+                  break;
+        case 'p':
+                  if (!strcmp(optarg, "win")) {
```
Comment (ivanhercaz) [inline i3lock.c]: Indentation error, should be like in `case 'd':`.

File: i3lock.c [hunk 7]
```diff
@@ -1004,105 +1004,148 @@ static void raise_loop(xcb_window_t window) {
...
+                      curs_choice = CURS_WIN;
+                  } else if (!strcmp(optarg, "default")) {
+                      curs_choice = CURS_DEFAULT;
+                  } else {
+                      errx(EXIT_FAILURE, "i3lock: Invalid pointer type given. Expected one of \"win\" or \"default\".");
+                  }
+                  break;
+        case 'e':
+                  ignore_empty_password = true;
```
Comment (ivanhercaz) [inline i3lock.c]: Indentation error, should be like in `case 'd':`.

File: i3lock.c [hunk 8]
```diff
@@ -1004,105 +1004,148 @@ static void raise_loop(xcb_window_t window) {
...
+                  break;
+        case 0:
+                  if (strcmp(longopts[longoptind].name, "debug") == 0)
```
Comment (ivanhercaz) [inline i3lock.c]: Indentation error, should be like in `case 'd':`.

File: i3lock.c [hunk 9]
```diff
@@ -1004,105 +1004,148 @@ static void raise_loop(xcb_window_t window) {
...
+                      debug_mode = true;
+                  else if (strcmp(longopts[longoptind].name, "raw") == 0)
+                      image_raw_format = strdup(optarg);
+                  break;
+        case 'f':
+                  show_failed_attempts = true;
```
Comment (ivanhercaz) [inline i3lock.c]: Indentation error, should be like in `case 'd':`.

File: i3lock.c [hunk 10]
```diff
@@ -1004,105 +1004,148 @@ static void raise_loop(xcb_window_t window) {
...
+                  break;
+        default:
+                  errx(EXIT_FAILURE, "Syntax: i3lock [-v] [-n] [-b] [-d] [-c color] [-u] [-p win|default]"
```
Comment (ivanhercaz) [inline i3lock.c]: Indentation error, should be like in `case 'd':`.

File: i3lock.c [hunk 11]
```diff
@@ -1004,105 +1004,148 @@ static void raise_loop(xcb_window_t window) {
...
+}
+
+int conf_file_rc() {
+
+    char buff[128] = {0};
+    char path[1024] = {0};
+    char *p = getenv("HOME");
+    int i = 0;
+
+    if(p == NULL)
```
Comment (ivanhercaz) [inline i3lock.c]: `if (p == NULL)`

Needs a space.

File: i3lock.c [hunk 12]
```diff
@@ -1004,105 +1004,148 @@ static void raise_loop(xcb_window_t window) {
...
+        return -ENOENT;
+
+    strcat(path, p);
+    strcat(path, "/.i3lockrc");
+    FILE *f = fopen(path, "r");
+
+    if(f == NULL)
```
Comment (ivanhercaz) [inline i3lock.c]: `if (f == NULL)`

Needs a space.

File: i3lock.c [hunk 13]
```diff
@@ -1004,105 +1004,148 @@ static void raise_loop(xcb_window_t window) {
...
+        return -ENOENT;
+
+    while(fgets(buff, sizeof buff, f) != NULL) {
```
Comment (ivanhercaz) [inline i3lock.c]: `while (fgets(buff, sizeof buff, f) != NULL)`

The `while` needs a space before the condition.

File: i3lock.c [hunk 14]
```diff
@@ -1004,105 +1004,148 @@ static void raise_loop(xcb_window_t window) {
...
+
```
Comment (ivanhercaz) [inline i3lock.c]: This newline should be removed.

File: i3lock.c [hunk 15]
```diff
@@ -1004,105 +1004,148 @@ static void raise_loop(xcb_window_t window) {
...
+        for(i=0; i<(sizeof longopts / sizeof *longopts) - 1; i++) {
```
Comment (ivanhercaz) [inline i3lock.c]: The `for` loop needs a space before `(i=0; i<(sizeof longopts / sizeof *longopts) - 1; i++)` and this line needs the next changes:
  - `i=0` » `i = 0`
  - `i<(sizeof longopts / sizeof *longopts)` » `i < (sizeof longopts / sizeof *longopts)`

File: i3lock.c [hunk 16]
```diff
@@ -1004,105 +1004,148 @@ static void raise_loop(xcb_window_t window) {
...
+            if(strncmp(longopts[i].name, buff, strlen(longopts[i].name)) == 0) {
```
Comment (ivanhercaz) [inline i3lock.c]: The `if` needs a space before `(strncmp(...`.

File: i3lock.c [hunk 17]
```diff
@@ -1004,105 +1004,148 @@ static void raise_loop(xcb_window_t window) {
...
+                if((p = strchr(buff, '\n')) != NULL)
```
Comment (ivanhercaz) [inline i3lock.c]: The `if` needs a space before `((p = strchr(...`.

File: i3lock.c [hunk 18]
```diff
@@ -1004,105 +1004,148 @@ static void raise_loop(xcb_window_t window) {
...
+                        *p = '\0';
+                set_opt(longopts[i].val, buff + strlen(longopts[i].name)+1, i);
```
Comment (ivanhercaz) [inline i3lock.c]: `strlen(longopts[i].name)+1` should be `strlen(longopts[i].name) + 1`.

**Comment by granquet** [comment]
@ivanhercaz : I'll update / cleanify the code and make it ready for the review / merge process as soon as @Airblader agrees the feature is indeed worth working on.

besides indentation, the code might be missing some checks here and there to make it robust and fail graciously whenever there's an issue.

**Comment by stapelberg** [comment]
We definitely shouldn’t add an ad-hoc config file parser; if we added anything, it would need to be using libconfuse just like i3status (rather than the custom parser i3 itself uses).

That said, I don’t think it’s worthwhile to have config file support for i3lock. One typically sets very few options, if any, and specifying them via flags works fine.

The cited use-case of interactively launching i3lock is better handled by starting a wrapper script or binding a key in i3 instead of using a program launcher.

**Comment by granquet** [comment]
> The cited use-case of interactively launching i3lock is better handled by starting a wrapper script or binding a key in i3 instead of using a program launcher.

let's agree to disagree on that one :)
I don't want another key binding to remember and I don't want to fiddle with my $PATH for one wrapper script since dmenu doesn't recognize aliases.

that's nothing I cannot live with... I just run i3lock with its defaults and it works, but that's not what I *really* want.

I'm still willing to work on the feature if there's any interest. feel free to re-open the merge request if you change your mind.

**Pull request closed without merging**